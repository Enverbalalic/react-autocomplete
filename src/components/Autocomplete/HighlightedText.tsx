import { useMemo } from "react";

export interface HighlightedTextProps {
    text: string,
    highlighted: string,
}

export interface SplitTextChunk {
    text: string,
    highlighted: boolean,
}

function splitText(fullText: string, highlighted: string): SplitTextChunk[] {
    if (highlighted.length === 0) {
        return [{
            text: fullText,
            highlighted: false,
        }];
    }

    let text = fullText;

    let chunks: SplitTextChunk[] = [];

    let highlightedIndex = text.toLowerCase().indexOf(highlighted.toLowerCase());

    while (highlightedIndex !== -1) {
        const highlightedEndIndex = highlightedIndex + highlighted.length;

        let before = text.substring(0, highlightedIndex);
        if (before.length > 0) {
            chunks.push({
                text: before,
                highlighted: false,
            });
        }

        let target = text.substring(highlightedIndex, highlightedEndIndex);
        let after = text.substring(highlightedEndIndex);

        chunks.push({
            text: target,
            highlighted: true,
        });

        text = after;
        highlightedIndex = text.toLowerCase().indexOf(highlighted.toLowerCase());

        if (highlightedIndex === -1 && after.length > 0) {
            chunks.push({
                text: after,
                highlighted: false,
            });
        }
    }

    return chunks;
}

function HighlightedText({ text, highlighted }: HighlightedTextProps) {
    const textChunks = useMemo(() => splitText(text, highlighted), [text, highlighted]);

    return (
        <>
            {textChunks.map((chunk, index) => {
                return (
                    <span key={index}
                          className={chunk.highlighted ? "eb__autocomplete-highlighted_text" : ""}>
                        {chunk.text}
                    </span>
                );
            })}
        </>
    );
}

export default HighlightedText;
