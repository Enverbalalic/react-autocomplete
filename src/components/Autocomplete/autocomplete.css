.eb-auto-complete {
    position: relative;
}

.eb-auto-complete .eb__autocomplete-input-container {
    display: flex;
    align-items: center;
}

.eb-auto-complete .eb__autocomplete-search-input {
    box-sizing: border-box;
    width: 100%;
    padding: 0.5rem;
    outline: none;
    border: 1px solid #999;
    border-radius: 6px;
}

.eb-auto-complete .eb__autocomplete-search-input:focus {
    padding: 0.5rem;
    outline: none;
    border: 1px solid #222;
    border-radius: 6px;
}

.eb-auto-complete .eb__autocomplete-error,
.eb-auto-complete .eb__loading-text,
.eb-auto-complete .eb__autocomplete-no-items-text {
    white-space: nowrap;
    font-size: 0.875rem;
    padding: 0.2rem 0.5rem;
}

.eb-auto-complete .eb__autocomplete-error {
    color: indianred;
}

.eb-auto-complete .eb__autocomplete-items-dropdown {
    position: absolute;
    max-height: 200px;
    border: 1px solid #999;
    border-radius: 6px;
    overflow: hidden;
    left: 0;
    right: 0;
    top: calc(100% + 0.5rem);
    display: flex;
    flex-direction: column;
}

.eb-auto-complete .eb__autocomplete-items {
    overflow: auto;
    height: 100%;
}


.eb-auto-complete .eb__autocomplete-item {
    padding: 0.5rem;
    cursor: pointer;
    transition: background-color 100ms ease-in;
}

.eb-auto-complete .eb__autocomplete-item .eb__autocomplete-highlighted_text {
    color: indianred;
}

.eb-auto-complete .eb__autocomplete-item:hover {
    background: rgba(0, 0, 0, 0.2);
}


.eb-auto-complete .eb__autocomplete-items::-webkit-scrollbar {
    width: 6px;
}

.eb-auto-complete .eb__autocomplete-items::-webkit-scrollbar-track {
    border-radius: 8px;
    background-color: #e7e7e7;
    border: 1px solid #cacaca;
}

.eb-auto-complete .eb__autocomplete-items::-webkit-scrollbar-thumb {
    border-radius: 8px;
    background-color: #999;
}
