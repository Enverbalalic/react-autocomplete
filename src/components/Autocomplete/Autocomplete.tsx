import { useCallback, useEffect, useState } from "react";

import "./autocomplete.css";
import HighlightedText from "src/components/Autocomplete/HighlightedText";
import useDebounce from "src/hooks/useDebounce";

export interface AutocompleteItem {
    id: number | string,
    name: number | string,
}

export interface AutocompleteProps<T extends AutocompleteItem> {
    /// Provided className will be appended to the root element of the component.
    className?: string,
    /// Will be called every time text in the input changes
    onChange: (search: string) => Promise<T[]>;
    /// Search input field placeholder
    placeholder?: string,
    /// Called when an item is clicked
    onSelect?: (item: T) => void,
    /// When a load error happens, this function gets called with the rejection reason, return a string to be displayed as the error
    onError?: (reason: any) => string,
    /// Custom text to show when there's no data
    noItemsText?: string,
    /// Custom text to show while we're loading data
    loadingText?: string,
    /// Custom delay between end of input and a onChange dispatch
    debounceDelay?: number,
}

export function Autocomplete<T extends AutocompleteItem>({
                                                             onSelect,
                                                             className,
                                                             placeholder = "Search...",
                                                             onChange,
                                                             onError,
                                                             debounceDelay = 500,
                                                             noItemsText = "No items",
                                                             loadingText = "Loading...",
                                                         }: AutocompleteProps<T>) {
    const [data, setData] = useState<T[]>([]);
    const [loading, setLoading] = useState<boolean>(false);
    const [error, setError] = useState<string | null>(null);
    const [text, setText] = useState("");
    const [isNoItems, setIsNoItems] = useState(false);
    const debouncedText = useDebounce(text, debounceDelay);

    useEffect(() => {
        if (debouncedText.length === 0) {
            setIsNoItems(false);
            setData([]);
            return;
        }

        setIsNoItems(false);
        setLoading(true);

        onChange(debouncedText).then(r => {
            setError(null);
            setData(r);

            if (r.length === 0) {
                setIsNoItems(true);
            }
        }).catch((error) => {
            setError(onError ? onError(error) : "We couldn't reach the search service. Please try again");
            setData([]);
        }).finally(() => {
            setLoading(false);
        });
    }, [debouncedText]);

    const onSelectItem = useCallback((item: T) => {
        if (!onSelect) return;

        setText("");
        setIsNoItems(false);
        setData([]);
        onSelect(item);
    }, [onSelect]);

    return (
        <div className={`eb-auto-complete${className ? ` ${className}` : ""}`}>
            <div className="eb__autocomplete-input-container">
                <input className="eb__autocomplete-search-input"
                       placeholder={placeholder}
                       value={text}
                       onChange={e => setText(e.target.value)}/>
                {loading && <div className="eb__loading-text">{loadingText}</div>}
                {error && <div className="eb__autocomplete-error">{error}</div>}
                {isNoItems && <div className="eb__autocomplete-no-items-text">{noItemsText}</div>}
            </div>

            {data.length > 0 && (
                <div className="eb__autocomplete-items-dropdown">
                    <div className="eb__autocomplete-items">
                        {data.map(item => {
                            return (
                                <div key={item.id}
                                     onClick={() => onSelectItem(item)}
                                     className="eb__autocomplete-item">
                                    <HighlightedText highlighted={debouncedText} text={item.name.toString()}/>
                                </div>
                            );
                        })}
                    </div>
                </div>
            )}
        </div>
    );
}

export default Autocomplete;
