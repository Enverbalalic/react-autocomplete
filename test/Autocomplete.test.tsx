import React from "react";
import renderer, { act, create } from "react-test-renderer";
import Autocomplete from "src/components/Autocomplete/Autocomplete";
import { describe, expect, vi } from "vitest";

function toJson(component: renderer.ReactTestRenderer) {
    const result = component.toJSON();
    expect(result).toBeDefined();
    expect(result).not.toBeInstanceOf(Array);
    return result as renderer.ReactTestRendererJSON;
}

describe("Autocomplete", () => {
    test("Renders correctly empty", () => {
        const component = renderer.create(
            <Autocomplete onChange={async (query) => {
                return [];
            }}></Autocomplete>,
        );

        let tree = toJson(component);
        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with items", async () => {
        vi.useFakeTimers();

        async function loadData() {
            return [{ id: 0, name: "Test" }, { id: 1, name: "Testing" }];
        }

        const loadDataMock = vi.fn().mockImplementation(loadData);

        let component = create(
            <Autocomplete onChange={loadDataMock}></Autocomplete>,
        );

        const inputEvent = { target: { value: "Test" } };

        await act(async () => {
            component.root.findByType("input").props.onChange(inputEvent);
            await vi.advanceTimersByTimeAsync(500);
        });

        vi.runAllTimers();

        expect(loadDataMock).toHaveBeenCalledOnce();

        expect(component.root.findByProps({ className: "eb__autocomplete-items" }).children.length).toEqual(2);

        const items = component.root.findAllByProps({ className: "eb__autocomplete-item" });

        expect(items[0].findByProps({ className: "eb__autocomplete-highlighted_text", children: "Test" })).not.toBeUndefined();

        expect(items[1].findByProps({ className: "eb__autocomplete-highlighted_text", children: "Test" })).not.toBeUndefined();
        expect(items[1].findByProps({ className: "", children: "ing" })).not.toBeUndefined();

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with no items", async () => {
        vi.useFakeTimers();

        async function loadData() {
            return [];
        }

        const loadDataMock = vi.fn().mockImplementation(loadData);

        let component = create(
            <Autocomplete onChange={loadDataMock}></Autocomplete>,
        );

        const inputEvent = { target: { value: "Test" } };

        await act(async () => {
            component.root.findByType("input").props.onChange(inputEvent);
            await vi.advanceTimersByTimeAsync(500);
        });

        vi.runAllTimers();

        expect(loadDataMock).toHaveBeenCalledOnce();

        expect(component.root.findByProps({ className: "eb__autocomplete-no-items-text" }).children).includes("No items");

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly while loading", async () => {
        vi.useFakeTimers();

        async function loadData() {
            await new Promise((resolve) => setTimeout(resolve, 500));
            throw new Error("Something went wrong");
        }

        const loadDataMock = vi.fn().mockImplementation(loadData);

        let component = create(
            <Autocomplete onChange={loadDataMock}></Autocomplete>,
        );

        const inputEvent = { target: { value: "Test" } };

        await act(async () => {
            component.root.findByType("input").props.onChange(inputEvent);
            await vi.advanceTimersByTimeAsync(500);
        });

        vi.runAllTimers();

        expect(loadDataMock).toHaveBeenCalledOnce();

        expect(component.root.findByProps({ className: "eb__loading-text" }).children).includes("Loading...");

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });


    test("Renders correctly with error promise", async () => {
        vi.useFakeTimers();

        async function loadData() {
            throw new Error("Something went wrong");
        }

        const loadDataMock = vi.fn().mockImplementation(loadData);

        let component = create(
            <Autocomplete onChange={loadDataMock}></Autocomplete>,
        );

        const inputEvent = { target: { value: "Test" } };

        await act(async () => {
            component.root.findByType("input").props.onChange(inputEvent);
            await vi.advanceTimersByTimeAsync(500);
        });

        vi.runAllTimers();

        expect(loadDataMock).toHaveBeenCalledOnce();

        expect(component.root.findByProps({ className: "eb__autocomplete-error" }).children).includes("We couldn't reach the search service. Please try again");

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with custom onError", async () => {
        vi.useFakeTimers();

        async function loadData() {
            throw new Error("Something went wrong");
        }

        const loadDataMock = vi.fn().mockImplementation(loadData);

        let component = create(
            <Autocomplete onError={() => "A custom error"} onChange={loadDataMock}></Autocomplete>,
        );

        const inputEvent = { target: { value: "Test" } };

        await act(async () => {
            component.root.findByType("input").props.onChange(inputEvent);
            await vi.advanceTimersByTimeAsync(500);
        });

        vi.runAllTimers();

        expect(loadDataMock).toHaveBeenCalledOnce();

        expect(component.root.findByProps({ className: "eb__autocomplete-error" }).children).includes("A custom error");

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });

    test("Renders correctly with custom className and placeholder", async () => {
        async function loadData() {
            return [];
        }

        let component = create(
            <Autocomplete className="custom-class"
                          placeholder="Custom placeholder" onChange={loadData}></Autocomplete>,
        );

        expect(component.root.findByType("div").props.className).includes("custom-class");
        expect(component.root.findByType("input").props.placeholder).toEqual("Custom placeholder");

        let tree = toJson(component!);

        expect(tree).toMatchSnapshot();
    });
});
