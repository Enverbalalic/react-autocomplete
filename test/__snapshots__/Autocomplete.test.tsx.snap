// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Autocomplete > Renders correctly empty 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value=""
    />
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly while loading 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value="Test"
    />
    <div
      className="eb__loading-text"
    >
      Loading...
    </div>
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly with custom className and placeholder 1`] = `
<div
  className="eb-auto-complete custom-class"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Custom placeholder"
      value=""
    />
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly with custom onError 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value="Test"
    />
    <div
      className="eb__autocomplete-error"
    >
      A custom error
    </div>
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly with error promise 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value="Test"
    />
    <div
      className="eb__autocomplete-error"
    >
      We couldn't reach the search service. Please try again
    </div>
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly with items 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value="Test"
    />
  </div>
  <div
    className="eb__autocomplete-items-dropdown"
  >
    <div
      className="eb__autocomplete-items"
    >
      <div
        className="eb__autocomplete-item"
        onClick={[Function]}
      >
        <span
          className="eb__autocomplete-highlighted_text"
        >
          Test
        </span>
      </div>
      <div
        className="eb__autocomplete-item"
        onClick={[Function]}
      >
        <span
          className="eb__autocomplete-highlighted_text"
        >
          Test
        </span>
        <span
          className=""
        >
          ing
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Autocomplete > Renders correctly with no items 1`] = `
<div
  className="eb-auto-complete"
>
  <div
    className="eb__autocomplete-input-container"
  >
    <input
      className="eb__autocomplete-search-input"
      onChange={[Function]}
      placeholder="Search..."
      value="Test"
    />
    <div
      className="eb__autocomplete-no-items-text"
    >
      No items
    </div>
  </div>
</div>
`;
